From 89103bd2f06dea6dc89276cd12dd78c72476b03f Mon Sep 17 00:00:00 2001
From: Alex Richardson <alexrichardson@google.com>
Date: Wed, 19 Jan 2022 09:53:41 +0000
Subject: [PATCH] [libc++] Fix GDB pretty printers when GDB uses Python 2.7

The gdb_pretty_printer_test.sh fails if GDB was built against Python 2.7
since Python 2 expects iterators to have a next() method rather than
using __next__. To make the pretty printers work with both Python 2 and 3
we can simply set next to __next__ in the iterator classes.

Python 2.7 support was removed in f46f93b4786377dd7ee704ef2beedadfe4f05adf,
so this partially reverts that commit. While Python 2.7 is EOL, it
appears there are still many GDB installations that are linked against
Python 2.7, so we may want to keep this tiny amount of compat code
around for a while longer.

Without this commit the tests fails with errors such as:
```
GDB printed:
   u"std::tuple containingTypeError: iter() returned non-iterator of type '_Children'\n"
Value should match:
   u'std::tuple containing = {[1] = 2, [2] = 3, [3] = 4}'
```

Reviewed By: #libc, ldionne

Differential Revision: https://reviews.llvm.org/D117470
---
 libcxx/utils/gdb/libcxx/printers.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/libcxx/utils/gdb/libcxx/printers.py b/libcxx/utils/gdb/libcxx/printers.py
index 2762091835d0..3f13b1f693c5 100644
--- a/libcxx/utils/gdb/libcxx/printers.py
+++ b/libcxx/utils/gdb/libcxx/printers.py
@@ -147,6 +147,8 @@ class StdTuplePrinter(object):
             self.count += 1
             return ("[%d]" % self.count, child)
 
+        next = __next__  # Needed for GDB built against Python 2.7.
+
     def __init__(self, val):
         self.val = val
 
@@ -356,6 +358,8 @@ class StdVectorPrinter(object):
                 self.offset = 0
             return ("[%d]" % self.count, outbit)
 
+        next = __next__  # Needed for GDB built against Python 2.7.
+
     class _VectorIterator(object):
         """Class to iterate over the non-bool vector's children."""
 
@@ -375,6 +379,8 @@ class StdVectorPrinter(object):
             self.item += 1
             return ("[%d]" % self.count, entry)
 
+        next = __next__  # Needed for GDB built against Python 2.7.
+
     def __init__(self, val):
         """Set val, length, capacity, and iterator for bool and normal vectors."""
         self.val = val
-- 
2.34.1.703.g22d0c6ccf7-goog

